src/github.com/matrix-org/dendrite/roomserver/input/membership.go

```diff
@@ -0,0 +1,223 @@
+		if err != nil {
+			return nil, err
+		}
+	}
+	if add != nil {
+		new, err = add.Membership()
+		if err != nil {
+			return nil, err
+		}
+	}
+	if old == new {
+		return updates, nil
+	}
+
+	mu, err := updater.MembershipUpdater(targetNID)
+	if err != nil {
+		return nil, err
+	}
+
+	return updateWithNewMembership(mu, new, add, updates)
+}
+
+func updateWithNewMembership(
+	mu types.MembershipUpdater, newMembership string, add *gomatrixserverlib.Event,
+	updates []api.OutputEvent,
+) ([]api.OutputEvent, error) {
+	switch newMembership {
+	case "invite":
+		needsSending, err := mu.SetToInvite(*add)
+		if err != nil {
+			return nil, err
+		}
+		if needsSending {
+			updates = appendInviteUpdate(updates, add)
+		}
+		return updates, nil
+	case "join":
+		if mu.IsJoin() {
+			return updates, nil
+		}
+		retired, err := mu.SetToJoin(add.Sender())
+		if err != nil {
+			return nil, err
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	case "leave", "ban":
+		if mu.IsLeave() {
+			return updates, nil
+		}
+		retired, err := mu.SetToLeave(add.Sender())
+		if err != nil {
+			mu.SetToLeave(add.Sender())
+		}
+		return appendRetireUpdates(updates, add, newMembership, retired), nil
+	default:
+		panic(fmt.Errorf(
+			"input: membership %q is not one of the allowed values", newMembership,
+		))
+	}
+}
+
+func appendInviteUpdate(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event,
+) []api.OutputEvent {
+	onie := api.OutputNewInviteEvent{
+		Event: *add,
+	}
+	return append(updates, api.OutputEvent{
+		Type:           api.OutputTypeNewInviteEvent,
+		NewInviteEvent: &onie,
+	})
+}
+
+func appendRetireUpdates(
+	updates []api.OutputEvent, add *gomatrixserverlib.Event, membership string, retired []string,
+) []api.OutputEvent {
+	for _, eventID := range retired {
+		orie := api.OutputRetireInviteEvent{
+			EventID:    eventID,
+			Membership: membership,
+		}
+		if add != nil {
+			orie.RetiredByEventID = add.EventID()
+		}
+		updates = append(updates, api.OutputEvent{
+			Type:              api.OutputTypeRetireInviteEvent,
+			RetireInviteEvent: &orie,
+		})
+	}
+	return updates
+}
+
+type stateChange struct {
+	removed types.StateEntry
+	added   types.StateEntry
+}
+
+func pairUpChanges(removed, added []types.StateEntry) []stateChange {
+	var ai int
+	var ri int
+	var result []stateChange
+	for {
+		switch {
+		case ai == len(added):
+			for _, s := range removed[ri:] {
+				result = append(result, stateChange{removed: s})
+			}
+			return result
+		case ri == len(removed):
+			for _, s := range added[ai:] {
+				result = append(result, stateChange{added: s})
+			}
+			return result
+		case added[ai].StateKeyTuple == removed[ri].StateKeyTuple:
+			result = append(result, stateChange{
+				removed: removed[ri],
+				added:   added[ai],
+			})
+			ai++
+			ri++
+		case added[ai].StateKeyTuple.LessThan(removed[ri].StateKeyTuple):
+			result = append(result, stateChange{added: added[ai]})
+			ai++
+		default:
```

Comment:
I don't like the fact that you've represented this as a switch/case. You really have 2 separate operations:
 - If we've reached the end, dump the remaining from the other slice into `result` then `return`.
 - Check if the `StateKeyTuple` is >, = or < and then append add/rm/both.

Separately, I don't understand why "if the state key tuple of added is less than the state key tuple of removed, append the added only". Why do we bias for lower entries? There's some significance I'm missing here, which you need to comment on.